### ISSUE TITLE
Use predictions from a joint model externally

### ISSUE THREAD

**Issue by OJay92** [CecileProust-Lima/lcmm#40]
Dear Professor Proust-Lima,

Is there a possibility to use a model, fitted on a training set (lets say **training_set**), outside of the global workspace where it has been fitted in. More specifically, I would like to apply my model **mod** on a test set **testing_set** which contains data similar to that of the training set but which is located somewhere else (lets say a computer/workspace of a colleague).. 

So the question is whether it is possible to compute predictions at another workspace just by transferring the fitted model and run it on a test set in another workspace that does not contain the original training set. 

Ive tried to experiment with this but unfortunately I received the logical error message which says:
`Error in eval(model$call$data) : object 'training_set' not found` 
after using this piece of code:
`predicted <- dynpred(mod, testing_set, landmark = c(unique(testing_set$week)),
                                     var.time = "week", horizon = c(1:12), draws = F)`

Thanks in advance,
OJ

**Comment by VivianePhilipps** [comment]
Hello,

yes, it is possible to use the dynpred function without keeping the training set in the workspace, but you have to plan it from the estmation of your model.
We added the 'returndata' argument in the Jointlcmm function (and in all other estimating functions) to do so.

An example with the data provided in the package :

library(lcmm)
d <- data_lcmm

m32 <- Jointlcmm(Ydep1~Time*X1,mixture=~Time,random=~Time,subject="ID",
     classmb=~X3,ng=2,survival=Surv(Tevent,Event)~X1+mixture(X2),
     hazard="3-quant-splines",hazardtype="PH",data=d,
     B = c(0.641, -0.6217, 0, 0, 0.5045, 0.8115, -0.4316, 0.7798, 0.1027, 
     0.7704, -0.0479, 10.4257, 11.2972, -2.5955, -0.5234, 1.4147, 
     -0.05, 0.9124, 0.0501, 0.2138, 1.5027), returndata=TRUE)

newd <- d[1:8,]
rm(d)

dynpred(m32,landmark=c(10,12),horizon=c(3,5,10),var.time="Time",
     fun.time=function(x){10*x},newdata=newd) # still works!

Best,

Viviane

**Comment by OJay92** [comment]
Dear Vivianne,

Indeed I noticed the functionality of the return data command which basically sends a duplicate of the training set along with the model.. But lets say one is in a situation where it is unsure whether it is desirable to "send" data along with a model (due to security/privacy issues).. would the use of the model still be possible.

I was more wondering why the model could not make its predictions without the presence of the original training set.. does it actually needs it before the "dynpred" part maybe due to statistical reasons (the random effects-part of the model for example).. ?

Thanks in advance and all the best,
OJ

**Comment by VivianePhilipps** [comment]
You are right, it may not be desirable to keep the data longer than necessary.
The dynpred function needs the data only to retrieve all observed levels of a factor included as a covariate in the model. Maybe we should think of an alternative way to store this information.

In the meanwhile, you can create a fake dataset including all the variables used in the model, making sure that all levels will appear for the possible factors.

For the above example :

data_fake <- data.frame(ID=1,Time=0,X1=0,X2=0,X3=0,Ydep1=0)
m32$data <- data_fake
dynpred(m32,landmark=c(10,12),horizon=c(3,5,10),var.time="Time",fun.time=function(x){10*x},newdata=newd) 

Viviane

**Comment by OJay92** [comment]
This works perfectly indeed.. i think this solution is pretty elegant as well!

Thanks a lot :)
